partitura/performance.py

```diff
@@ -280,6 +284,172 @@ def adjust_offsets_w_sustain(
+            "pitch",
+            "note_on",
+            "note_off",
+            "velocity",
+            "track",
+            "channel",
+            "sound_off",
+            "note_on_tick",
+            "note_off_tick",
+        ]
+
+    def __str__(self):
+        return f"PerformedNote: {self['id']}"
+
+    def __eq__(self, other):
+        if not isinstance(PerformedNote):
+            return False
+        if not self.keys() == other.keys():
+            return False
+        return np.all(
+            np.array([self[k] == other[k] for k in self.keys() if k in other.keys()])
+        )
+
+    def keys(self):
+        return self.pnote_dict.keys()
+
+    def get(self, key, default=None):
+        return self.pnote_dict.get(key, default)
+
+    def __hash__(self):
+        return hash(self["id"])
+
+    def __lt__(self, other):
+        return self["note_on"] < other["note_on"]
+
+    def __le__(self, other):
+        return self["note_on"] <= other["note_on"]
+
+    def __gt__(self, other):
+        return self["note_on"] > other["note_on"]
+
+    def __ge__(self, other):
+        return self["note_on"] >= other["note_on"]
+
+    def __getitem__(self, key):
+        return self.pnote_dict.get(key, None)
+
+    def __setitem__(self, key, value):
+        if key not in self._accepted_keys:
+            raise KeyError(f"Key {key} not accepted for PerformedNote")
+        self.pnote_dict[key] = value
+        self._validate_values(key)
+
+    def __delitem__(self, key):
+        raise KeyError("Cannot delete items from PerformedNote")
+
+    def __iter__(self):
+        return iter(self.keys())
+
+    def __len__(self):
+        return len(self.keys())
+
+    def __contains__(self, key):
+        return key in self.keys()
+
+    def copy(self):
+        return PerformedNote(self.pnote_dict.copy())
+
+    def _validate_values(self, key=None):
+        if key is None:
+            keys = self.keys()
+        else:
+            keys = [key]
+
+        for key in keys:
+            if key == "pitch":
+                self._validate_pitch()
+            elif key == "note_on":
+                self._validate_note_on()
+            elif key == "note_off":
+                self._validate_note_off()
+            elif key == "velocity":
+                self._validate_velocity()
+            elif key == "sound_off":
+                self._validate_sound_off()
+
+    def _validate_sound_off(self):
+        if self.get("note_off", -1) < 0:
+            return
+        if self.pnote_dict["sound_off"] < self.pnote_dict["note_off"]:
+            raise ValueError(f"sound_off must be greater or equal to note_off")
+
+    def _validate_note_on(self):
+        if self.pnote_dict["note_on"] < 0:
+            raise ValueError(
+                f"Note on value provided is invalid, must be greater than or equal to 0"
+            )
+
+    def _validate_note_off(self):
+        if self.pnote_dict.get("note_on", -1) < 0:
+            return
+        if (
```

Comment:
maybe check for negative offsets also in this case?

Reply:
I think I already do in the if clause.

Reply:
yes, but only after this check. if the `note`_on`( or `note_on_tick`) is below 0, the second test is never reached and you can set an arbitrary offset.
another minor thing is that there is no upper limit for the onset setters, so there can be negative durations by setting higher onsets than offsets. 
this is only minor stuff, let me know if you want to fix it, it's fine to leave it as is too.